File: packages/core/src/canonical.ts
/**
 * Deterministic JSON serialization: object keys sorted recursively so the same
 * logical record always hashes to the same bytes regardless of insertion order.
 * `undefined` values are dropped (matching JSON.stringify semantics for objects).
 */
export function canonicalJson(value: unknown): string {
  return JSON.stringify(sortValue(value));
}

function sortValue(value: unknown): unknown {
  if (Array.isArray(value)) return value.map(sortValue);
  if (value !== null && typeof value === "object") {
    const record = value as Record<string, unknown>;
    const out: Record<string, unknown> = {};
    for (const key of Object.keys(record).sort()) {
      const v = record[key];
      if (v !== undefined) out[key] = sortValue(v);
    }
    return out;
  }
  return value;
}

File: src/toolgate/core/canonical.py
import json
from typing import Any


def canonical_json(value: Any) -> str:
    """Deterministic JSON: keys sorted recursively, compact separators.

    The same logical record always hashes to the same bytes regardless of
    insertion order. `None` values are preserved; callers drop absent fields
    before canonicalization (pydantic `exclude_none`).
    """
    return json.dumps(value, sort_keys=True, separators=(",", ":"), ensure_ascii=False)
